File: android/src/main/java/dragon3/save/SaveData.java
package dragon3.save;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import dragon3.common.Body;
import dragon3.stage.StageStatus;

@lombok.Data
public class SaveData implements Serializable, Cloneable {

	private static final long serialVersionUID = 761813354588427943L;

	private boolean reverse;
	private boolean allClear;
	private int score;
	private int turn;
	private int dead;
	private int kill;
	private int item;
	private int escape;
	private int save;
	private long playTime;
	private String playerName = "Player";
	
	private List<Body> bodyList = new ArrayList<>();
	private Map<String, StageStatus> stageStatusMap = new HashMap<>();

	public SaveData copy() {
		try {
			return (SaveData) this.clone();
		} catch (CloneNotSupportedException e) {
			throw new RuntimeException(e);
		}
	}

	public void countTurn() {
		turn++;
	}
	public void countDead() {
		dead++;
	}
	public void countKill() {
		kill++;
	}
	public void countItem() {
		item++;
	}
	public void countEscape() {
		escape++;
	}
	public void countSave() {
		save++;
	}
	public void addTime(long n) {
		playTime += n;
	}
	
	public StageStatus getStageStatus(String stageId) {
		if (!stageStatusMap.containsKey(stageId)) {
			StageStatus stageStatus = new StageStatus();
			stageStatus.setStageId(stageId);
			stageStatusMap.put(stageId, stageStatus);
		}
		return stageStatusMap.get(stageId);
	}
	
	public int getStarNum(String stageId) {
		return getStageStatus(stageId).getStar();
	}
	
	public void countStarNum(String stageId) {
		StageStatus status = getStageStatus(stageId);
		status.setStar(status.getStar() + 1);
	}
	
	public void setOpened(String stageId, boolean flag) {
		StageStatus status = getStageStatus(stageId);
		status.setOpened(flag);
	}
	
	public int sumStars() {
		int sum = 0;
		for (StageStatus status : stageStatusMap.values()) {
			sum += status.getStar();
		}
		return sum;
	}
}

File: core/src/main/java/dragon3/save/SaveData.kt
package dragon3.save

import java.io.Serializable
import java.util.ArrayList
import java.util.HashMap

import dragon3.common.Body
import dragon3.stage.StageStatus

class SaveData : Serializable, Cloneable {

    var reverse: Boolean = false
    var allClear: Boolean = false
    var score: Int = 0
    var turn: Int = 0
    var dead: Int = 0
    var kill: Int = 0
    var item: Int = 0
    var escape: Int = 0
    var save: Int = 0
    var playTime: Long = 0
    var playerName = "Player"

    var bodyList :List<Body> = ArrayList<Body>()
    var stageStatusMap = HashMap<String, StageStatus>()

    fun copy(): SaveData {
        try {
            return this.clone() as SaveData
        } catch (e: CloneNotSupportedException) {
            throw RuntimeException(e)
        }

    }

    fun countTurn() {
        turn++
    }

    fun countDead() {
        dead++
    }

    fun countKill() {
        kill++
    }

    fun countItem() {
        item++
    }

    fun countEscape() {
        escape++
    }

    fun countSave() {
        save++
    }

    fun addTime(n: Long) {
        playTime += n
    }

    fun getStageStatus(stageId: String): StageStatus {
        if (!stageStatusMap.containsKey(stageId)) {
            val stageStatus = StageStatus()
            stageStatus.stageId = stageId
            stageStatusMap.put(stageId, stageStatus)
        }
        return stageStatusMap[stageId]!!
    }

    fun getStarNum(stageId: String): Int {
        return getStageStatus(stageId).star
    }

    fun countStarNum(stageId: String) {
        val status = getStageStatus(stageId)
        status.star = status.star + 1
    }

    fun setOpened(stageId: String, flag: Boolean) {
        val status = getStageStatus(stageId)
        status.opened = flag
    }

    fun sumStars(): Int {
        var sum = 0
        for (status in stageStatusMap.values) {
            sum += status.star
        }
        return sum
    }

}
